Al final del [texto anterior](3.0_Switch.md), mencionamos que cada secuencia de instrucciones después de una etiqueta `case` o `default` debía terminar con una declaración `break` o `return`. 

Pronto exploraremos por qué, además de algunas *sutilezas* sobre el alcance (*scope*) de las declaraciones `switch` que pueden llegar a causarnos problemas si no las tratamos con cuidado.

## Caída

Cuando una expresión condicional evaluada en `switch` coincide con una etiqueta `case` o cae en la etiqueta `default`, la ejecución empieza en la instrucción inmediata posterior a la etiqueta correspondiente. La ejecución continuará hasta que se satisfaga alguna de las siguientes condiciones de término:
1. Llegamos al final del bloque `switch` (o sea, el corchete de cierre).
2. Un `break` o `return` nos saca del bloque `switch` o de la función.
3. Algo más interrumpe a nuestro programa (ej.: Se nos congela la computadora, alguien divide entre cero y el universo explota, etc.)

Nótese que la existencia de otras etiquetas (sean `case`o `default`) **no** es una condición terminal. Entonces, sin un `break` o `return`, la ejecución del programa se desbordará en casos posteriores.

Veamos un ejemplo:
```
#include <iostream>

int main()
{
    switch (2)
    {
    case 1: // No coincide
        std::cout << 1 << '\n'; // Saltado
    case 2: // Match!
        std::cout << 2 << '\n'; // Coincide, la ejecución empieza aquí
    case 3:
        std::cout << 3 << '\n'; // Esto también se ejecuta
    case 4:
        std::cout << 4 << '\n'; // Esto también se ejecuta
    default:
        std::cout << 5 << '\n'; // Esto también se ejecuta
    }

    return 0;
}
```
En el supuesto de que estemos usando [la configuración sugerida para VSCode](../Caja_de_herramientas/Configurar_el_compilador.md) (y claramente: VSCode), el programa anterior ni siquiera compilará y nos dará como error algo similar a `this statement may fall through [-Werror=implicit-fallthrough=]g`. 

En caso de que tengamos una configuración más permisiva, el output será:
```
2
3
4
5
```

Probablemente, esta no era nuestra intención. Cuando la ejecución fluye desde una declaración debajo de una etiqueta a declaraciones debajo de etiquetas subsecuentes, decimos que hubo **caída** o falla (*fallthrough*).

Como las *caídas* no suelen ser intencionales, muchos compiladores las señalarán como un fallo o advertencia.

### El atributo [[fallthrough]]

En caso de que intencionalmente escribamos código con *caídas*, es una práctica común escribir algún comentario que le indique a otros programadores (o a nosotros mismos 3 meses en el futuro que ya nisiquiera recordemos qué hacía el programa) que dicho comportamiento de nuestro programa es intencional. Lastimosamente, el compilador no sabe leer comentarios, así que de todos modos nos lo señalará como un error (y en configuraciones restrictivas, ni siquiera compilará nuestro código).

